### PR TITLE
Advanced logging config options in google_compute_subnetwork

### PR DESCRIPTION
**Pull request by dustin-decker**
Added custom metadata fields and filter expressions to subnetwork flow log configuration. Fixes https://github.com/hashicorp/terraform-provider-google/issues/6500

Signed-off-by: Dustin Decker <dustin.decker@getcruise.com>

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Added custom metadata fields and filter expressions to `google_compute_subnetwork` flow log configuration
```

### CODE REVIEW

**Comment by danawillow** [comment]
This fixes https://github.com/terraform-providers/terraform-provider-google/issues/6500,  right? Can you put that in the PR description?

I'll take a look once the CLA is resolved, but my comments are almost certainly going to be "can you add a test for this?"

**Comment by dustin-decker** [comment]
Yes, it should resolve https://github.com/terraform-providers/terraform-provider-google/issues/6500

Thanks for mentioning, I didn't see the issue before. I'll look at adding a test.

**Comment by dustin-decker** [comment]
@googlebot I signed it!

**Comment by dustin-decker** [comment]
Added terraform test

File: third_party/terraform/tests/resource_compute_subnetwork_test.go [hunk 1]
```diff
@@ -577,6 +589,32 @@ resource "google_compute_subnetwork" "network-with-flow-logs" {
...
+
+resource "google_compute_subnetwork" "network-with-flow-logs" {
+  name          = "%s"
+  ip_cidr_range = "10.0.0.0/16"
+  region        = "us-central1"
+  network       = google_compute_network.custom-test.self_link
+  log_config {
+    aggregation_interval = "INTERVAL_30_SEC"
+    flow_sampling        = 0.8
+	metadata             = "CUSTOM_METADATA"
```
Comment (danawillow) [inline third_party/terraform/tests/resource_compute_subnetwork_test.go:607]: watch out for tabs vs spaces

File: products/compute/api.yaml [hunk 2]
```diff
@@ -12966,14 +12966,24 @@ objects:
             values:
               - :EXCLUDE_ALL_METADATA
               - :INCLUDE_ALL_METADATA
+              - :CUSTOM_METADATA
             default_value: :INCLUDE_ALL_METADATA
+          - !ruby/object:Api::Type::Array
+            name: 'metadataFields'
+            description: |
+              Can only be specified if VPC flow logs for this subnetwork is enabled and "metadata" was set to CUSTOM_METADATA.
```
Comment (danawillow) [inline products/compute/api.yaml:12974]:
```suggestion
              Can only be specified if VPC flow logs for this subnetwork is enabled and "metadata" is set to CUSTOM_METADATA.
```

File: products/compute/api.yaml [hunk 3]
```diff
@@ -12966,14 +12966,24 @@ objects:
             values:
               - :EXCLUDE_ALL_METADATA
               - :INCLUDE_ALL_METADATA
+              - :CUSTOM_METADATA
             default_value: :INCLUDE_ALL_METADATA
+          - !ruby/object:Api::Type::Array
```
Comment (danawillow) [inline products/compute/api.yaml:12971]: Does order matter here? If not, I'd recommend marking this as `is_set: true` in terraform.yaml

File: third_party/terraform/tests/resource_compute_subnetwork_test.go (hunk 1 above)
Comment (dustin-decker) [reply third_party/terraform/tests/resource_compute_subnetwork_test.go]: fixed

File: products/compute/api.yaml (hunk 3 above)
Comment (dustin-decker) [reply products/compute/api.yaml:12971]: Set this value to make it treated as a set

File: products/compute/api.yaml (hunk 2 above)
Comment (dustin-decker) [reply products/compute/api.yaml]: Changed

**Comment by dustin-decker** [comment]
I've incorporated your suggestions and added the custom expanders and flatteners. Thanks for the pointers, there is a lot to get familiar with in this repo. Tests are passing on my machine.

File: templates/terraform/custom_expand/subnetwork_log_config.go.erb [hunk 4]
```diff
@@ -35,6 +35,14 @@ func expand<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d T
 	transformed["aggregationInterval"] = original["aggregation_interval"]
 	transformed["flowSampling"] = original["flow_sampling"]
 	transformed["metadata"] = original["metadata"]
+	
+	// make it JSON marshallable
+	transformed["metadataFields"] = original["metadata_fields"].(*schema.Set).List()
+
+	// allow a null default (must eval to true otherwise)
+	if val := reflect.ValueOf(original["filter_expr"]); val.IsValid() && !isEmptyValue(val) {
```
Comment (danawillow) [inline templates/terraform/custom_expand/subnetwork_log_config.go.erb:43]: I'm  not 100% sure, but I think this is going to mean that `filter_expr` will not be able to be updated to an empty value after its been set (it depends on how the API handles unset fields in the PATCH request). Can you check?

Comment (dustin-decker) [reply templates/terraform/custom_expand/subnetwork_log_config.go.erb:43]: I've added a test that I hope covers unsetting this and the metadata fields. The API did not accept an empty value, but would accept a unset value.

Comment (dustin-decker) [reply templates/terraform/custom_expand/subnetwork_log_config.go.erb:43]: Oops.. test is failing. I'll take a closer look.

Comment (dustin-decker) [reply templates/terraform/custom_expand/subnetwork_log_config.go.erb:43]: I'm having some trouble with this. 

If it's unset, I think it is not updating.

If it sends a empty string it gets this error:

```
Error: Error creating Subnetwork: googleapi: Error 400: Invalid value for field 'resource.logConfig.filterExpr': ''. ERROR: VPC Flow Log filter expression:1:0: mismatched input '<EOF>' expecting {'[', '{', '(', '.', '-', '!', 'true', 'false', 'null', NUM_FLOAT, NUM_INT, NUM_UINT, STRING, BYTES, IDENTIFIER}
    | 
    | ^
ERROR: VPC Flow Log filter expression:1:0: Filter expression does not evaluate to a boolean
    | 
    | ^, invalid

    on /tmp/tf-test941063018/main.tf line 7:
    (source code not available)
```

Is the solution to send the filter expression string `true` when it's unset? That feels wrong.

Comment (danawillow) [reply templates/terraform/custom_expand/subnetwork_log_config.go.erb:43]: That does feel wrong. If you have the exact request/response from GCP (which you can get by running the test with env var TF_LOG=DEBUG) I can help look into it.

**Comment by danawillow** [comment]
I played around with this a bit today, and I think it's a bug in the API. I filed a bug against the team (b/162771344 for googlers), and will keep you posted on progress! If it ends up taking too long to resolve, we could also just mark the field as not updatable in order to get the functionality in.

**Comment by dustin-decker** [comment]
Thanks Dana! Sorry I let this sit so long. I suspected the same and put off trying to thoroughly prove it. I'll be happy to pick it back up when there's a path forward.

**Comment by danawillow** [comment]
All right, I heard back from the API team and their recommendation was just to default to "true", since that's equivalent to leaving it empty. I'm now trying to find out whether it's possible to set `metadataFields` to empty as well, since that seemed to fail when I ran the tests locally.

**Comment by danawillow** [comment]
Ok, they said that for metadataFields clearing it is equivalent to setting metadata to EXCLUDE_ALL_METADATA. I haven't actually tried it in the API yet to see if that clears the field or not. Let me know if you want me to take over any of this! I'd love to get it in soon.

**Comment by dustin-decker** [comment]
I'm busy for the next couple of weeks so if you'd like to take it the rest of the way I'd be grateful. We're looking forward to this feature too.

**Pull request merged**